### PR TITLE
ddl: fix partition schemadiff operation

### LINKED ISSUE #22819: ddl: truncate partition results in incorrect schema diff
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
/* init */ drop table if exists t;
/* init */ create table t (id int primary key, v int) partition by range (id) (partition p0 values less than (4), partition p_rest values less than maxvalue);
/* init */ insert into t values (1, 10), (3, 30), (5, 50);
/* txn */ begin;
/* txn */ update t set id = 2 where id = 5;
/* ddl */ alter table t truncate partition p0; 
/* txn */ commit; -- This is expected to fail

```
It seems the schema diff is not applied properly for the `truncate` statement. 
The result of `Builder.ApplyDiff` is an emptry array containing no related table ids, which is unexpected.

### 2. What did you expect to see? (Required)
Error is reported like schema has changed.

### 3. What did you see instead (Required)
The commit succeeded.

### 4. What is your TiDB version? (Required)
master branch `9cb9b69b61a76e9cc3f8cfa27f86635c2496cf16` commit hash

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Issue Number: close #22819

Problem Summary: #19991 added the support for partition-level control of truncating partition. #20361 added support of incremental placement rules. #20361 changed the behavior target from old partition IDs to partitions that needs to update placement rules.

The solution:

1. IDs are composed of a list of `(oldID, -1)` or `(oldID, newID)`.
2. For `(oldID,-1)`, use `oldID` to finish #19991 and drop the placement rules
3. For `(oldID,newID)`, doing 2, then fetch new rules using `newID`

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

refer to #22819

### Release note 

- No release note

### CODE REVIEW

**Comment by bb7133** [comment]
PTAL @djshow832 @AilinKid

**Comment by cfzjywxk** [comment]
@bb7133 
The `ApplyDiff` is in the critical path, and it's a bit difficult to understand its logic. Maybe we should refactor this part and add some defensive code as well as more unit-tests for it as it's closely related to correctness.

**Comment by xhebox** [comment]
@cfzjywxk I agree. But compatibility is a big concern. Either we made a compromise of compatibility, or we must keep the old and new `ApplyDiff` system at the same time.

The original semantics of `SchemaDiff` is not general at all. I can't come up with a refactor without breaking the compatibility.

File: infoschema/builder.go [hunk 1]
```diff
@@ -146,11 +146,13 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) ([]int64, erro
...
-
 				b.applyPlacementDelete(placement.GroupID(opt.OldTableID))
-				err := b.applyPlacementUpdate(placement.GroupID(opt.TableID))
-				if err != nil {
-					return nil, errors.Trace(err)
+				if opt.TableID != -1 {
+					// (!= -1) => fetching new placement rules from PD
+					err := b.applyPlacementUpdate(placement.GroupID(opt.TableID))
+					if err != nil {
+						return nil, errors.Trace(err)
```
Comment (AilinKid) [inline infoschema/builder.go:154]: you can put the test case of the issue into the unit test.

**Review by AilinKid — COMMENTED** [review]
LGTM

File: infoschema/builder.go (hunk 1 above)
Comment (AilinKid) [reply infoschema/builder.go:154]: maybe we can add it in the automated test ( be tested in every CI round)

Comment (xhebox) [reply infoschema/builder.go:154]: > seems applyDiff logic can't be tested in the unit test

We can, but need a more cleaner API and in-process TiDB. I mentioned that in my WIP sharing.

> maybe we can add it in the automated test

Yep, automated-test can handle it. This problem, or the original issue is not closely related to placement rules. It seems that we did not test DDL/transaction things in CI. Do we run those tests before 5.0-rc? 5.0-rc should have the bug already.

**Review by djshow832 — COMMENTED** [review]
Test cases?

**Comment by xhebox** [comment]
> Test cases?

> seems applyDiff logic can't be tested in the unit test, we need two at least TiDB node. -- AilinKid

File: ddl/partition.go [hunk 2]
```diff
@@ -1102,18 +1102,18 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
...
-				yoldIDs = append(yoldIDs, oldID)
 				newIDs = append(newIDs, newPartitions[i].ID)
 				bundles = append(bundles, placement.BuildPlacementDropBundle(oldID))
 				bundles = append(bundles, placement.BuildPlacementCopyBundle(oldBundle, newPartitions[i].ID))
+			} else {
+				newIDs = append(newIDs, -1)
 			}
 		}
-		job.CtxVars = []interface{}{yoldIDs, newIDs}
+		job.CtxVars = []interface{}{oldIDs, newIDs}
```
Comment (djshow832) [inline ddl/partition.go:1116]: How about this:
oldIDs = old partition IDs
newIDs = new partition IDs
In applyDiff, always drop placement rules for old partitions, and always fetch placement rules for new partitions, no matter it exists or not.

**Comment by djshow832** [comment]
> > Test cases?
> 
> > seems applyDiff logic can't be tested in the unit test, we need two at least TiDB node. -- AilinKid
> 
> One TiDB instance could test it, though. No one will bother to write unit tests under the current API. It is more practical to add some integration tests elsewhere.

I tested on my notebook with only one TiDB instance and no TiKV instance, the bug still happens.

File: ddl/partition.go (hunk 2 above)
Comment (xhebox) [reply ddl/partition.go:1116]: So if you truncate 10 partitions, you got 10 rtt of  TiDB-PD.

**Comment by xhebox** [comment]
> I tested on my notebook with only one TiDB instance and no TiKV instance, the bug still happens.

Strange, I can not reproduce.

**Comment by djshow832** [comment]
> > I tested on my notebook with only one TiDB instance and no TiKV instance, the bug still happens.
> 
> Strange, I can not reproduce.
> 
> **NOTE**: you need another new session to run DDL. wjhuang2016 notified that to me. It is actually a problem of two sessions of concurrent operations. In one session, DDL should terminate the transaction, then there is no error.

Yes, of course. I used another session to run the DDL. Then the DML transaction commits successfully.

**Comment by xhebox** [comment]
> Yes, of course. I used another session to run the DDL. Then the DML transaction commits successfully.

`if d.infoHandle != nil && d.infoHandle.IsValid() {` Then the only possible case is that it did not run into this loop.

File: ddl/partition.go [hunk 3]
```diff
@@ -1099,29 +1099,34 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
...
 
 		err = infosync.PutRuleBundles(nil, bundles)
 		if err != nil {
 			job.State = model.JobStateCancelled
 			return ver, errors.Wrapf(err, "failed to notify PD the placement rules")
 		}
+	} else {
+		for i := range oldIDs {
+			newIDs[i] = -1
```
Comment (djshow832) [inline ddl/partition.go:1125]: `newIDs` should not only depend on the implementation of placement rules. By iteral, it means the new partition IDs.

Comment (xhebox) [reply ddl/partition.go:1125]: I knew. If you think always fetching rules are acceptable, or adding new fields to SchemaDiff is acceptable, we can do that.

File: infoschema/builder.go (hunk 1 above)
Comment (AilinKid) [reply infoschema/builder.go:154]: `seems that we did not test DDL/transaction things in CI`, yes, multi-node sync logic for DDL hasn't been tested. I think we should, how do you think @bb7133 @djshow832

File: ddl/partition.go (hunk 3 above)
Comment (djshow832) [reply ddl/partition.go:1125]: I think it's fine to always fetch rules from PD.

Comment (xhebox) [reply ddl/partition.go:1125]: Just one more time: is it acceptable to introduce, in the worst case, `2*x` RTT of TiDB/PD latency for the truncation of `x` partitions, even if the user is not using placement rules at all? Give me a `Yes`, and I will modify to always fetch rules.

Comment (djshow832) [reply ddl/partition.go:1125]: Why 2*x RTT? Deleting placement rules from the cache doesn't need fetching rules.

Comment (xhebox) [reply ddl/partition.go:1125]: Oh, yes. So there is constant factor `2`.

Comment (djshow832) [reply ddl/partition.go:1125]: Then I think it's fine. Please modify to always fetch rules.

Comment (xhebox) [reply ddl/partition.go]: Done

**Review by djshow832 — COMMENTED** [review]
/LGTM

**Comment by djshow832** [comment]
LGTM

**Pull request merged**